### ISSUE TITLE
Feedback on implementation of verbose output feature

### ISSUE THREAD

**Issue by lcowell** [exercism/cli#163]
I noticed that the `verbose` flag isn't actually being used anywhere in the code and I think it would be really useful to log information like what url was requested, and what json response came back. Being able to see this information  would give me (and other developers) a more concrete understanding of what's happening behind the scenes and make it easier to develop features and debug interactions.

I think what I want to do is divide output up and use `fmt` for data that we want the user to see under normal operations and use the `log` package for verbose output. Why I like this approach is that I can set the default logger via `log.SetOutput(ioutil.Discard)` and that means a call like `log.Println("hello")` will output nothing.

The downside of this is that many lines where we're using log to output information to the user would need to be changed. I didn't want to make that sort of sweeping change without checking in. The other options is that I could add our own logger package that could handle different levels of logging output.

The other thing I'm look at is how to handle the verbose flag. The verbose flag is only available through `cli.Context`, so it seems like we'd need to get the flag information at the beginning of each `cmd` function. It would be nice to avoid this sort of repetition and handle the `verbose` flag before we delegate to the command.

**Comment by kytrinyx** [comment]
Yeah, the verbose thing was a hack that I put in specifically for a debug release at one point, and then took it out, because it should be done properly if we do it.

I think the distinction of fmt for output to the user and log for verbose and fatal errors would be fine. I don't see a reason to create our own logger at this point.

For handling the verbose flag, we could set it as a variable on the cmd package before starting. Not a great practice in general, but here we're not likely to have trouble with race conditions because of it.

**Comment by Tonkpils** [comment]
This is a good idea. One nice thing that we could do is wrapping the `log` package so that we can insert our own messages when a Fatal error happens. Maybe having a message logged that guides the user to submit an issue.

**Comment by kytrinyx** [comment]
Yeah, wrapping log.Fatal would be great.

**Comment by lcowell** [comment]
I wonder if rather than simply wrapping the stdlib log package, it should provide a slightly different interface. I'm wondering if we should be able to set a log level at the package level and add methods like `log.Info`, `log.Infof`, `log.Warn` and `log.Crit`. Those methods would only produce output if the log level is high enough. Is this overkill? Would it be enough to have debug logging on or off ? I could see the most verbose mode being too much for normal circumstances, but good if you wanted to see everything that's going back and forth between client and server.

**Comment by Tonkpils** [comment]
I like the idea of the interface. However, we have a couple of different things going on here though. One is the information that is displayed to the user which can be success or error information. Second, we have extra logging that would only be shown when the verbose/debug option is on or off. We have either verbose or not. Errors however, would always/should always need to be displayed and are not necessarily tied to verbosity. 

How about some examples.

*Verbose turned off*
```
Info> Submitted exercise successfully
```
```
Error> Failed to submit exercise
```
*Verbose turned on*
```
Debug> Contacting API...
Debug> Parsing JSON... 
Info> Submitted exercise successfully
```
```
Debug> Contacting API...
Debug> Error from API Status: 500 
Error> Failed to submit exercise
```

**Comment by lcowell** [comment]
Thanks for clarifying @Tonkpils . I've pushed something up to my fork here:
https://github.com/exercism/cli/compare/master...lcowell:logger_output

Does this seem like it's on the right track ? I'd probably add a generic `Print` function and add functions like `Errorf` so we can also use formatting.

**Comment by Tonkpils** [comment]
That looks good to me. :+1:

**Comment by lcowell** [comment]
I'm just taking another pass at this and I was wondering what you think about multiline output. Say we have a message like this:

<pre>
Submitted bob in ruby.
Your submission can be found online at http://url.com/path

To get the next exercise, run "exercism fetch" again.
</pre>

The current behaviour would be to only prepend the first line with "Info". I think this might make the output look inconsistent. I think I could prepend each line with `Info`, but it raises the question of whether that prefix just becomes noise to the user.

**Comment by kytrinyx** [comment]
The "Info" stuff seems like noise to me. I don't think we should have log-like output unless we turn on the --verbose flag. It's kind of more like UI/UX than logging at this point.

**Comment by Tonkpils** [comment]
:+1: to not having the "Info" and just adding log-like output for --verbose.

**Comment by kytrinyx** [comment]
I think this is no longer relevant, since #248 went into master. Would either of you confirm and close if this is true?
